crates/oxc_linter/src/service.rs

```diff
@@ -250,26 +251,44 @@ impl Runtime {
             }
         };
 
-        let sources = PartialLoader::parse(ext, &source_text);
-        let is_processed_by_partial_loader = sources.is_some();
-        let sources = sources
+        let sources = PartialLoader::parse(ext, &source_text)
             .unwrap_or_else(|| vec![JavaScriptSource::partial(&source_text, source_type, 0)]);
 
         if sources.is_empty() {
             self.ignore_path(path);
             return;
         }
 
-        for JavaScriptSource { source_text, source_type, .. } in sources {
+        // If there are fixes, we will accumulate all of them and write to the file at the end.
+        // This means we do not write multiple times to the same file if there are multiple sources
+        // in the same file (for example, multiple scripts in an `.astro` file).
+        let mut new_source_text = Cow::from(&source_text);
+        // This is used to keep track of the difference in the offset when fixing the source.
+        // Otherwise, spans will not take into account fixed code that comes before them in the source.
+        let mut fix_offset: i32 = 0;
+
+        for source in sources {
             let allocator = Allocator::default();
-            let mut messages =
-                self.process_source(path, &allocator, source_text, source_type, true, tx_error);
+            let mut messages = self.process_source(
+                path,
+                &allocator,
+                source.source_text,
+                source_type,
+                true,
+                tx_error,
+            );
 
-            // TODO: Span is wrong, ban this feature for file process by `PartialLoader`.
-            if !is_processed_by_partial_loader && self.linter.options().fix.is_some() {
-                let fix_result = Fixer::new(source_text, messages).fix();
+            if self.linter.options().fix.is_some() {
+                let fix_result = Fixer::new(source.source_text, messages).fix();
                 if fix_result.fixed {
-                    fs::write(path, fix_result.fixed_code.as_bytes()).unwrap();
+                    // write to file, replacing only the changed part
+                    let start = source.start.saturating_add_signed(fix_offset) as usize;
+                    let end = start + source.source_text.len();
+                    new_source_text.to_mut().replace_range(start..end, &fix_result.fixed_code);
+                    let old_code_len = u32::try_from(source.source_text.len()).unwrap();
+                    let new_code_len = u32::try_from(fix_result.fixed_code.len()).unwrap();
+                    fix_offset += i32::try_from(new_code_len).unwrap();
+                    fix_offset -= i32::try_from(old_code_len).unwrap();
```

Comment:
I would love any ideas on how to avoid all of this number casting here. the main issue is that `start..end` has to be a range of `usize`, but `source.start` is `u32`, while `fix_offset` needs to potentially be negative (as source files can be smaller after applying fixes), so it becomes `i32`.

Reply:
Just cast them with `as`, we assume source length is under u32.

Reply:
https://github.com/oxc-project/oxc/blob/51a78d594640a069db53940bdb838fb420ea698c/crates/oxc_parser/src/lib.rs#L22-L33

Reply:
right, I am often forgetting about that, I'll update this to just cast and ignore truncation